### DISCUSSION TITLE
Could oxiri::Iri be a trait instead of an implementation?

### DISCUSSION THREAD

**Discussion by JervenBolleman** [oxigraph/oxigraph#822] [Ideas]
For @heringerp project it would be helpful if Iri was a trait. In this way we could make a lazy IRI implementation that only allocates the strings representation when rendering. At the same time we could also hide some extra information in lazy Iri objects (pointers to backend data structures).

**Comment by Tpt** [comment]
Hey! Yes, lazy string representation would be great! However, I am not sure to understand where you want to have them (sorry, I am missing some context on "heringerp project"). In Oxigraph `oxiri::Iri` is only used to validate IRIs and resolve relative ones. `NamedNode` (and `NamedNodeRef`) are the structs used in the APIs to represent RDF IRIs. The `Store` implementation providing SPARQL does not uses these types but rather an internal `EncodingTerm` struct that provides dictionary support and only builds the `NamedNode` object at the boundaries.

Moving a struct `S` to a trait `T` as significant cost on the codebase verbosity because either the trait needs to be wrapped into a dynamic pointer like (`Box<dyn T>`, `Arc<dyn T>` or `&dyn T`) or the concrete implementation of `T` to be used needs to be injected using generics (like `struct Triple<ConcreteIriType>`). It's why I have avoided to do it to keep the APIs a bit simpler. But I would be happy to change this decision if there are worth it use cases in the places where it can be useful.

**Reply by JervenBolleman** [reply]
@Tpt thanks for the reply, I understand things better now. Yes, we don't actually need to make oxiri:Iri lazy inside the SPARQL against genome handlegraph datastructures as an limited alternative backend store implementation. We can work with the infrastructure that is present already.

**Reply by Tpt** [reply]
Thank you! So, I guess you want to reuse Oxigraph SPARQL evaluator on top of handlegraph? Currently the evaluator is sadly quite relying on the `EncodedTerm` struct, I would love to make it more generic (help in this space very welcome!). I have also a plan to investigate if I can reuse [DataFusion](https://arrow.apache.org/datafusion/) for the evaluation (doing proper query optimization and fast vectorised execution is a lot of work, I would love to be able to piggyback on something working established)

**Reply by JervenBolleman** [reply]
Yes, @heringerp is reusing the evaluation logic of Oxigraph [see this branch](https://github.com/heringerp/oxigraph/tree/dev). Reusing data fusion for the logical plans sounds great.
